File: ts/src/instrumentation/processor.ts
/**
 * LLMSpanProcessor — converts OpenInference span attributes to typed LLMSpan objects.
 */

import type { Context } from "@opentelemetry/api";
import { SpanStatusCode } from "@opentelemetry/api";
import type { ReadableSpan, SpanProcessor } from "@opentelemetry/sdk-trace-base";

import type { DeliveryQueue } from "./queue";
import type {
  AssistantMessage,
  LLMSpan,
  Message,
  ToolCall,
  ToolDefinition,
} from "./spans";
import {
  createAssistantMessage,
  createImageContent,
  createSystemMessage,
  createTextContent,
  createToolResultMessage,
  createUserMessage,
} from "./spans";

// ── Helper functions ─────────────────────────────────────────────────────────

function inferProvider(modelName: string): string {
  const lower = modelName.toLowerCase();
  if (lower.includes("gpt") || lower.includes("o1") || lower.includes("o3")) {
    return "openai";
  }
  if (lower.includes("claude")) return "anthropic";
  if (lower.includes("gemini")) return "google";
  if (lower.includes("command") || lower.includes("coral")) return "cohere";
  if (
    lower.includes("llama") ||
    lower.includes("mixtral") ||
    lower.includes("mistral")
  ) {
    return "meta";
  }
  return "unknown";
}

function parseJson(raw: string): Record<string, unknown> {
  try {
    const result = JSON.parse(raw);
    if (typeof result === "object" && result !== null && !Array.isArray(result)) {
      return result as Record<string, unknown>;
    }
    return {};
  } catch {
    return {};
  }
}

function toFloatOrNull(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  const n = Number(value);
  return Number.isNaN(n) ? null : n;
}

function toIntOrNull(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  const n = parseInt(String(value), 10);
  return Number.isNaN(n) ? null : n;
}

type Attrs = Record<string, unknown>;

function parseContentParts(
  attrs: Attrs,
  prefix: string
): Array<{ type: "text"; text: string } | { type: "image"; url: string; detail: string | null }> {
  const parts: Array<
    { type: "text"; text: string } | { type: "image"; url: string; detail: string | null }
  > = [];

  let j = 0;
  while (true) {
    const typeKey = `${prefix}.contents.${j}.message_content.type`;
    const contentType = attrs[typeKey];
    if (contentType === undefined || contentType === null) break;

    if (contentType === "text") {
      const textKey = `${prefix}.contents.${j}.message_content.text`;
      const text = String(attrs[textKey] ?? "");
      parts.push(createTextContent(text));
    } else if (contentType === "image") {
      const urlKey = `${prefix}.contents.${j}.message_content.image.url.url`;
      const detailKey = `${prefix}.contents.${j}.message_content.image.url.detail`;
      const url = String(attrs[urlKey] ?? "");
      const detailRaw = attrs[detailKey];
      const detail = detailRaw != null ? String(detailRaw) : null;
      parts.push(createImageContent(url, detail));
    }

    j++;
  }

  if (parts.length === 0) {
    const contentKey = `${prefix}.content`;
    const contentRaw = attrs[contentKey];
    if (contentRaw != null) {
      parts.push(createTextContent(String(contentRaw)));
    }
  }

  return parts;
}

function parseToolCalls(attrs: Attrs, prefix: string): ToolCall[] {
  const toolCalls: ToolCall[] = [];
  let j = 0;
  while (true) {
    const nameKey = `${prefix}.tool_calls.${j}.tool_call.function.name`;
    const name = attrs[nameKey];
    if (name === undefined || name === null) break;

    const argsKey = `${prefix}.tool_calls.${j}.tool_call.function.arguments`;
    const argsRaw = attrs[argsKey];
    let args: Record<string, unknown>;
    if (typeof argsRaw === "string") {
      try {
        args = JSON.parse(argsRaw);
      } catch {
        args = { _raw: argsRaw };
      }
    } else if (typeof argsRaw === "object" && argsRaw !== null) {
      args = argsRaw as Record<string, unknown>;
    } else {
      args = {};
    }

    const idKey = `${prefix}.tool_calls.${j}.tool_call.id`;
    const callIdRaw = attrs[idKey];
    const callId = callIdRaw != null ? String(callIdRaw) : null;

    toolCalls.push({ name: String(name), arguments: args, id: callId });
    j++;
  }

  return toolCalls;
}

function parseInputMessages(attrs: Attrs): Message[] {
  const messages: Message[] = [];
  let i = 0;
  while (true) {
    const prefix = `llm.input_messages.${i}.message`;
    const roleKey = `${prefix}.role`;
    const role = attrs[roleKey];
    if (role === undefined || role === null) break;

    const roleStr = String(role).toLowerCase();

    if (roleStr === "system") {
      const contentKey = `${prefix}.content`;
      const content = String(attrs[contentKey] ?? "");
      messages.push(createSystemMessage(content));
    } else if (roleStr === "user") {
      const parts = parseContentParts(attrs, prefix);
      messages.push(createUserMessage(parts));
    } else if (roleStr === "assistant") {
      const parts = parseContentParts(attrs, prefix);
      const toolCalls = parseToolCalls(attrs, prefix);
      messages.push(createAssistantMessage({ content: parts, toolCalls }));
    } else if (roleStr === "tool") {
      const contentKey = `${prefix}.content`;
      const content = String(attrs[contentKey] ?? "");
      const toolCallIdRaw = attrs[`${prefix}.tool_call_id`];
      const toolCallId = toolCallIdRaw != null ? String(toolCallIdRaw) : null;
      const toolNameRaw = attrs[`${prefix}.name`];
      const toolName = toolNameRaw != null ? String(toolNameRaw) : null;
      messages.push(
        createToolResultMessage({ content, toolCallId, toolName })
      );
    }

    i++;
  }
  return messages;
}

function parseOutputMessages(attrs: Attrs): AssistantMessage[] {
  const messages: AssistantMessage[] = [];
  let i = 0;
  while (true) {
    const prefix = `llm.output_messages.${i}.message`;
    const roleKey = `${prefix}.role`;
    const role = attrs[roleKey];
    if (role === undefined || role === null) break;

    const parts = parseContentParts(attrs, prefix);
    const toolCalls = parseToolCalls(attrs, prefix);
    const finishReasonRaw = attrs[`${prefix}.finish_reason`];
    const finishReason =
      finishReasonRaw != null ? String(finishReasonRaw) : null;

    messages.push(
      createAssistantMessage({ content: parts, toolCalls, finishReason })
    );
    i++;
  }
  return messages;
}

function parseToolDefinitions(attrs: Attrs): ToolDefinition[] {
  const tools: ToolDefinition[] = [];
  let i = 0;
  while (true) {
    const nameKey = `llm.tools.${i}.tool.name`;
    const name = attrs[nameKey];
    if (name === undefined || name === null) break;

    const descRaw = attrs[`llm.tools.${i}.tool.description`];
    const description = descRaw != null ? String(descRaw) : null;

    const schemaRaw = attrs[`llm.tools.${i}.tool.json_schema`];
    let parameters: Record<string, unknown> | null = null;
    if (typeof schemaRaw === "string") {
      const parsed = parseJson(schemaRaw);
      parameters = Object.keys(parsed).length > 0 ? parsed : null;
    } else if (typeof schemaRaw === "object" && schemaRaw !== null) {
      parameters = schemaRaw as Record<string, unknown>;
    }

    tools.push({ name: String(name), description, parameters });
    i++;
  }
  return tools;
}

const NULL_SPAN_ID = "0000000000000000";

// ── Processor ────────────────────────────────────────────────────────────────

/**
 * OTel SpanProcessor that converts OpenInference LLM spans to typed
 * LLMSpan objects and submits them to the delivery queue.
 */
export class LLMSpanProcessor implements SpanProcessor {
  private readonly _deliveryQueue: DeliveryQueue;

  constructor(deliveryQueue: DeliveryQueue) {
    this._deliveryQueue = deliveryQueue;
  }

  onStart(_span: ReadableSpan, _parentContext?: Context): void {
    // No-op — we only process completed spans.
  }

  onEnd(span: ReadableSpan): void {
    try {
      const attrs: Attrs = span.attributes
        ? { ...span.attributes }
        : {};

      const spanKind = attrs["openinference.span.kind"];
      if (spanKind !== "LLM" && spanKind !== "EMBEDDING") {
        return;
      }

      const llmSpan = this._buildLlmSpan(span, attrs, String(spanKind));
      this._deliveryQueue.submit(llmSpan);
    } catch {
      // Never raise from onEnd
    }
  }

  shutdown(): Promise<void> {
    return Promise.resolve();
  }

  forceFlush(): Promise<void> {
    return this._deliveryQueue.flush().then(() => undefined);
  }

  private _buildLlmSpan(
    span: ReadableSpan,
    attrs: Attrs,
    spanKind: string
  ): LLMSpan {
    // Identity / timing
    const ctx = span.spanContext();
    const spanId = ctx.spanId;
    const traceId = ctx.traceId;
    const parentSpanId = span.parentSpanId && span.parentSpanId !== NULL_SPAN_ID
      ? span.parentSpanId
      : null;

    const startHr = span.startTime;
    const endHr = span.endTime;
    const startMs = startHr[0] * 1000 + startHr[1] / 1e6;
    const endMs = endHr[0] * 1000 + endHr[1] / 1e6;
    const startedAt = new Date(startMs);
    const endedAt = new Date(endMs);
    const durationMs = endMs - startMs;

    // Provider / model
    const requestModel = String(
      attrs["llm.model_name"] ?? attrs["gen_ai.request.model"] ?? ""
    );
    const responseModelRaw = attrs["gen_ai.response.model"];
    const responseModel =
      responseModelRaw != null ? String(responseModelRaw) : null;
    const provider =
      String(attrs["gen_ai.system"] ?? "") || inferProvider(requestModel);
    const operation = spanKind === "EMBEDDING" ? "embedding" : "chat";

    // Token usage
    const inputTokens = Number(attrs["llm.token_count.prompt"] ?? 0);
    const outputTokens = Number(attrs["llm.token_count.completion"] ?? 0);
    const cacheReadTokens = Number(
      attrs["llm.token_count.cache_read"] ?? 0
    );
    const cacheCreationTokens = Number(
      attrs["llm.token_count.cache_creation"] ?? 0
    );

    // Request parameters
    const params = parseJson(
      String(attrs["llm.invocation_parameters"] ?? "{}")
    );
    const requestTemperature = toFloatOrNull(params["temperature"]);
    const requestMaxTokens = toIntOrNull(
      params["max_tokens"] ?? params["max_completion_tokens"]
    );
    const requestTopP = toFloatOrNull(params["top_p"]);

    // Response / error
    const responseIdRaw =
      attrs["llm.response_id"] ?? attrs["gen_ai.response.id"];
    const responseId = responseIdRaw != null ? String(responseIdRaw) : null;
    const outputTypeRaw = attrs["gen_ai.output.type"];
    const outputType = outputTypeRaw != null ? String(outputTypeRaw) : null;
    const errorTypeRaw = attrs["error.type"];
    let errorType: string | null = null;
    if (errorTypeRaw != null) {
      errorType = String(errorTypeRaw);
    } else if (
      span.status &&
      span.status.code === SpanStatusCode.ERROR
    ) {
      errorType = "error";
    }

    // Messages
    const inputMessages = parseInputMessages(attrs);
    const outputMessages = parseOutputMessages(attrs);
    const finishReasons = outputMessages
      .map((m) => m.finishReason)
      .filter((r): r is string => r !== null);

    // Tool definitions
    const toolDefinitions = parseToolDefinitions(attrs);

    return {
      spanId,
      traceId,
      parentSpanId,
      startedAt,
      endedAt,
      durationMs,
      operation,
      provider,
      requestModel,
      responseModel,
      inputTokens,
      outputTokens,
      cacheReadTokens,
      cacheCreationTokens,
      requestTemperature,
      requestMaxTokens,
      requestTopP,
      finishReasons,
      responseId,
      outputType,
      errorType,
      inputMessages,
      outputMessages,
      toolDefinitions,
    };
  }
}

File: pixie/instrumentation/processor.py
"""LLMSpanProcessor — converts OpenInference span attributes to LLMSpan."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

from opentelemetry.sdk.trace import ReadableSpan, SpanProcessor
from opentelemetry.trace import StatusCode

from .queue import _DeliveryQueue
from .spans import (
    AssistantMessage,
    ImageContent,
    LLMSpan,
    Message,
    SystemMessage,
    TextContent,
    ToolCall,
    ToolDefinition,
    ToolResultMessage,
    UserMessage,
)


class LLMSpanProcessor(SpanProcessor):
    """OTel SpanProcessor that converts OpenInference LLM spans to typed LLMSpan objects."""

    def __init__(self, delivery_queue: _DeliveryQueue) -> None:
        self._delivery_queue = delivery_queue

    def on_start(self, span: Any, parent_context: Any = None) -> None:
        """No-op — we only process completed spans."""

    def on_end(self, span: ReadableSpan) -> None:
        """Convert completed OpenInference LLM spans to LLMSpan and submit."""
        try:
            attrs = dict(span.attributes) if span.attributes else {}

            # Only process LLM spans
            span_kind = attrs.get("openinference.span.kind")
            if span_kind not in ("LLM", "EMBEDDING"):
                return

            llm_span = self._build_llm_span(span, attrs, str(span_kind))
            self._delivery_queue.submit(llm_span)

            # Write to trace file if a writer is active
            try:
                from dataclasses import asdict

                from pixie.instrumentation.trace_writer import get_trace_writer

                writer = get_trace_writer()
                if writer is not None:
                    writer.write_llm_span(asdict(llm_span))
            except Exception:
                pass
        except Exception:
            pass  # Never raise from on_end

    def on_shutdown(self) -> None:
        """No-op."""

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        """Flush the delivery queue."""
        return self._delivery_queue.flush(timeout_seconds=timeout_millis / 1000)

    def _build_llm_span(
        self,
        span: ReadableSpan,
        attrs: dict[str, Any],
        span_kind: str,
    ) -> LLMSpan:
        """Build a typed LLMSpan from raw OTel span and attributes."""
        # ── Identity / timing
        ctx = span.context
        if ctx is None:
            raise ValueError("No span context")
        span_id = format(ctx.span_id, "016x")
        trace_id = format(ctx.trace_id, "032x")
        parent_span_id = format(span.parent.span_id, "016x") if span.parent else None

        start_ns = span.start_time or 0
        end_ns = span.end_time or 0
        started_at = datetime.fromtimestamp(start_ns / 1e9, tz=timezone.utc)
        ended_at = datetime.fromtimestamp(end_ns / 1e9, tz=timezone.utc)
        duration_ms = (end_ns - start_ns) / 1e6

        # ── Provider / model
        request_model = str(attrs.get("llm.model_name") or attrs.get("gen_ai.request.model", ""))
        response_model_raw = attrs.get("gen_ai.response.model")
        response_model = str(response_model_raw) if response_model_raw is not None else None
        provider = str(attrs.get("gen_ai.system", "")) or _infer_provider(request_model)
        operation = "embedding" if span_kind == "EMBEDDING" else "chat"

        # ── Token usage
        input_tokens = int(attrs.get("llm.token_count.prompt", 0))
        output_tokens = int(attrs.get("llm.token_count.completion", 0))
        cache_read_tokens = int(attrs.get("llm.token_count.cache_read", 0))
        cache_creation_tokens = int(attrs.get("llm.token_count.cache_creation", 0))

        # ── Request parameters
        params = _parse_json(str(attrs.get("llm.invocation_parameters", "{}")))
        request_temperature = _to_float_or_none(params.get("temperature"))
        request_max_tokens = _to_int_or_none(
            params.get("max_tokens") or params.get("max_completion_tokens")
        )
        request_top_p = _to_float_or_none(params.get("top_p"))

        # ── Response / error
        response_id_raw = attrs.get("llm.response_id") or attrs.get("gen_ai.response.id")
        response_id = str(response_id_raw) if response_id_raw is not None else None
        output_type_raw = attrs.get("gen_ai.output.type")
        output_type = str(output_type_raw) if output_type_raw is not None else None
        error_type_raw = attrs.get("error.type")
        if error_type_raw is not None:
            error_type: str | None = str(error_type_raw)
        elif span.status and span.status.status_code == StatusCode.ERROR:
            error_type = "error"
        else:
            error_type = None

        # ── Messages
        input_messages = _parse_input_messages(attrs)
        output_messages = _parse_output_messages(attrs)
        finish_reasons = tuple(msg.finish_reason for msg in output_messages if msg.finish_reason)

        # ── Tool definitions
        tool_definitions = _parse_tool_definitions(attrs)

        return LLMSpan(
            span_id=span_id,
            trace_id=trace_id,
            parent_span_id=parent_span_id,
            started_at=started_at,
            ended_at=ended_at,
            duration_ms=duration_ms,
            operation=operation,
            provider=provider,
            request_model=request_model,
            response_model=response_model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cache_read_tokens=cache_read_tokens,
            cache_creation_tokens=cache_creation_tokens,
            request_temperature=request_temperature,
            request_max_tokens=request_max_tokens,
            request_top_p=request_top_p,
            finish_reasons=finish_reasons,
            response_id=response_id,
            output_type=output_type,
            error_type=error_type,
            input_messages=tuple(input_messages),
            output_messages=tuple(output_messages),
            tool_definitions=tuple(tool_definitions),
        )


# ── Helper functions ──────────────────────────────────────────────────────────


def _infer_provider(model_name: str) -> str:
    """Infer the LLM provider from the model name."""
    lower = model_name.lower()
    if "gpt" in lower or "o1" in lower or "o3" in lower:
        return "openai"
    if "claude" in lower:
        return "anthropic"
    if "gemini" in lower:
        return "google"
    if "command" in lower or "coral" in lower:
        return "cohere"
    if "llama" in lower or "mixtral" in lower or "mistral" in lower:
        return "meta"
    return "unknown"


def _parse_json(raw: str) -> dict[str, Any]:
    """Parse JSON safely, returning empty dict on failure."""
    try:
        result = json.loads(raw)
        if isinstance(result, dict):
            return result
        return {}
    except (json.JSONDecodeError, TypeError, ValueError):
        return {}


def _to_float_or_none(value: Any) -> float | None:
    """Convert value to float or return None."""
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _to_int_or_none(value: Any) -> int | None:
    """Convert value to int or return None."""
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _parse_content_parts(
    attrs: dict[str, Any], prefix: str
) -> tuple[TextContent | ImageContent, ...]:
    """Parse multimodal content parts from OpenInference indexed attributes.

    Falls back to plain .content string as a single TextContent.
    """
    parts: list[TextContent | ImageContent] = []
    j = 0
    while True:
        type_key = f"{prefix}.contents.{j}.message_content.type"
        content_type = attrs.get(type_key)
        if content_type is None:
            break

        if content_type == "text":
            text_key = f"{prefix}.contents.{j}.message_content.text"
            text = str(attrs.get(text_key, ""))
            parts.append(TextContent(text=text))
        elif content_type == "image":
            url_key = f"{prefix}.contents.{j}.message_content.image.url.url"
            detail_key = f"{prefix}.contents.{j}.message_content.image.url.detail"
            url = str(attrs.get(url_key, ""))
            detail_raw = attrs.get(detail_key)
            detail = str(detail_raw) if detail_raw is not None else None
            parts.append(ImageContent(url=url, detail=detail))

        j += 1

    if not parts:
        # Fall back to plain .content string
        content_key = f"{prefix}.content"
        content_raw = attrs.get(content_key)
        if content_raw is not None:
            parts.append(TextContent(text=str(content_raw)))

    return tuple(parts)


def _parse_tool_calls(attrs: dict[str, Any], prefix: str) -> tuple[ToolCall, ...]:
    """Parse tool calls from OpenInference indexed attributes."""
    tool_calls: list[ToolCall] = []
    j = 0
    while True:
        name_key = f"{prefix}.tool_calls.{j}.tool_call.function.name"
        name = attrs.get(name_key)
        if name is None:
            break

        args_key = f"{prefix}.tool_calls.{j}.tool_call.function.arguments"
        args_raw = attrs.get(args_key)
        if isinstance(args_raw, str):
            try:
                arguments = json.loads(args_raw)
            except json.JSONDecodeError:
                arguments = {"_raw": args_raw}
        elif isinstance(args_raw, dict):
            arguments = args_raw
        else:
            arguments = {}

        id_key = f"{prefix}.tool_calls.{j}.tool_call.id"
        call_id_raw = attrs.get(id_key)
        call_id = str(call_id_raw) if call_id_raw is not None else None

        tool_calls.append(ToolCall(name=str(name), arguments=arguments, id=call_id))
        j += 1

    return tuple(tool_calls)


def _parse_input_messages(attrs: dict[str, Any]) -> list[Message]:
    """Parse input messages from OpenInference indexed span attributes."""
    messages: list[Message] = []
    i = 0
    while True:
        prefix = f"llm.input_messages.{i}.message"
        role_key = f"{prefix}.role"
        role = attrs.get(role_key)
        if role is None:
            break

        role = str(role).lower()

        if role == "system":
            content_key = f"{prefix}.content"
            content = str(attrs.get(content_key, ""))
            messages.append(SystemMessage(content=content))
        elif role == "user":
            parts = _parse_content_parts(attrs, prefix)
            messages.append(UserMessage(content=parts))
        elif role == "assistant":
            parts = _parse_content_parts(attrs, prefix)
            tool_calls = _parse_tool_calls(attrs, prefix)
            messages.append(AssistantMessage(content=parts, tool_calls=tool_calls))
        elif role == "tool":
            content_key = f"{prefix}.content"
            content = str(attrs.get(content_key, ""))
            tool_call_id_raw = attrs.get(f"{prefix}.tool_call_id")
            tool_call_id = str(tool_call_id_raw) if tool_call_id_raw is not None else None
            tool_name_raw = attrs.get(f"{prefix}.name")
            tool_name = str(tool_name_raw) if tool_name_raw is not None else None
            messages.append(
                ToolResultMessage(content=content, tool_call_id=tool_call_id, tool_name=tool_name)
            )

        i += 1

    return messages


def _parse_output_messages(attrs: dict[str, Any]) -> list[AssistantMessage]:
    """Parse output messages from OpenInference indexed span attributes."""
    messages: list[AssistantMessage] = []
    i = 0
    while True:
        prefix = f"llm.output_messages.{i}.message"
        role_key = f"{prefix}.role"
        role = attrs.get(role_key)
        if role is None:
            break

        parts = _parse_content_parts(attrs, prefix)
        tool_calls = _parse_tool_calls(attrs, prefix)

        # finish_reason is per-message in OpenInference
        finish_reason_raw = attrs.get(f"{prefix}.finish_reason")
        finish_reason = str(finish_reason_raw) if finish_reason_raw is not None else None

        messages.append(
            AssistantMessage(
                content=parts,
                tool_calls=tool_calls,
                finish_reason=finish_reason,
            )
        )
        i += 1

    return messages


def _parse_tool_definitions(attrs: dict[str, Any]) -> list[ToolDefinition]:
    """Parse tool definitions from OpenInference indexed span attributes."""
    tools: list[ToolDefinition] = []
    i = 0
    while True:
        name_key = f"llm.tools.{i}.tool.name"
        name = attrs.get(name_key)
        if name is None:
            break

        desc_raw = attrs.get(f"llm.tools.{i}.tool.description")
        description = str(desc_raw) if desc_raw is not None else None

        schema_raw = attrs.get(f"llm.tools.{i}.tool.json_schema")
        if isinstance(schema_raw, str):
            parameters = _parse_json(schema_raw) or None
        elif isinstance(schema_raw, dict):
            parameters = schema_raw
        else:
            parameters = None

        tools.append(
            ToolDefinition(name=str(name), description=description, parameters=parameters)
        )
        i += 1

    return tools
